### ISSUE TITLE
Model ensembling with yolov8

### ISSUE THREAD

**Issue by whatisslove11** [ultralytics/ultralytics#3036]
### Search before asking

- [x] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Good morning!
I wanna train several custom yolov8 models, but I haven't found any guides for yolov8 for this theme.

For yolov5 I found this guide (https://docs.ultralytics.com/yolov5/tutorials/model_ensembling/), but I still have some questions.

1. How can we create model ensemble wihtout CLI (I wanna use python)?
2. How can we start training process in python for ensemble on custom dataset?
3. How can we save our models after training and how we can test ensemble of custom models? 

If possible, if there is also a solution using python for all tasks, please provde it.
Thanks!

**Comment by glenn-jocher** [comment]
@whatisslove11 hello! 

Glad to hear you're interested in using YOLOv8 for model ensembling. 

1. To create model ensembles without using the CLI, you can use the `models.load_ensemble` function, which loads a list of models, and then use the `ensemble` method to generate detections from all of them at once.
2. To start the training process in Python for custom datasets, we recommend using the `models.train_single` function to train each individual model in the ensemble on your custom dataset, and then loading the resulting weights using `models.load` to create your ensemble list.
3. To save your models after training, you can use the `models.save` function, and to test your ensemble of custom models, you can simply use the `ensemble` method to generate detections. 

We hope this helps! Let us know if you have any further questions.

**Comment by whatisslove11** [comment]
@glenn-jocher  thank u so much for your answer!

**Comment by whatisslove11** [comment]
@glenn-jocher hi!
I am so sorry, my futher question will ve very stupid, but... I haven't find any information about `models.load_ensemble`, `models.train_single` and etc. Have you got any tutorials for this functions? Can you drop it gere, please? Thanks a lot and sorry for such a stupid question

**Comment by rumjie** [comment]
@glenn-jocher Hi glenn, I am facing the same problem with @whatisslove11. It is hard to find models.load_ensemble in ultralytics, I guess you mentioned the function in yolov5. Could you please let me know if I have to write manual code to ensemble yolov8 models for now?

**Comment by glenn-jocher** [comment]
@rumjie hello,

Thank you for reaching out. I apologize for any confusion caused. As of now, there is no built-in `models.load_ensemble` function in YOLOv8. To ensemble YOLOv8 models, you would need to manually load and combine the individual models yourself.

You can load each YOLOv8 model using the `models.load` function and then use the `ensemble` method to generate detections from all the models simultaneously.

To ensemble YOLOv8 models for now, you will need to write manual code to load and combine the models.

Please let me know if you have any further questions.

Thank you!

**Comment by jaideep11061982** [comment]
@glenn-jocher  can you please given an eg. 
I understand we can load the models.  but nt getting how to ensemble ?

**Comment by glenn-jocher** [comment]
@jaideep11061982 to ensemble models in YOLOv8, you have to load each individual trained model and then pass an image or a set of images through each model one by one. Each model will output predictions which are basically bounding boxes with class scores. 

To ensemble the predictions, you generally have two methods:
- **Voting-based method** (also known as majority voting): Each model votes for each class and the class with the most votes is chosen.
- **Average method**: Each model outputs the probabilities for each class and the final output class is the one with the highest average probability.

Note that ensembling may not always lead to better performance. It highly depends on how diverse and accurate your individual models are. If your models are very similar or one model's performance is far worse than the others, ensembling might not help. 

Always evaluate your ensemble model using a validation set to see if it actually improves the performance. 

I hope this helps! Let me know if you have more questions.

**Comment by rakshit-bhardwaj** [comment]
@glenn-jocher I am trying to ensemble 3 yolov8 models trained on visible, infrared and thermal datasets. Can you explain how the bounding boxes will be affected? I was wondering if it is possible to create combined bounding boxes from the individual bounding boxes by taking weighted average of them and with their contribution proportional to their confidence. Also please tell me if this is a stupid idea or not. and if it is not, how will I input the image to the ensembled model? will the input image be some sort of combination of images of the 3 spectrums?

**Comment by glenn-jocher** [comment]
@rakshit-bhardwaj It's certainly not a stupid idea; it's quite innovative to ensemble models trained on different spectrums like visible, infrared, and thermal. This approach can leverage the unique advantages of each spectrum to improve detection under various conditions.

When ensembling these models, you can indeed combine the individual bounding boxes by taking a weighted average based on confidence scores. This technique is often referred to as **bbox fusion** or **confidence-weighted averaging** and can help in creating a more robust bounding box that takes into account the "opinions" of each model according to their confidence levels.

The logic behind this is that if one model is very confident about a detection (say, the thermal model detects a warm object in low-light conditions), this detection should carry more weight in the final bounding box coordinates.

As for inputting the image, if the models are trained separately on different spectrum images, you will need to run inference on each model separately with their respective spectrum images. Then you collect the predictions from each model and ensemble them as discussed above. 

The input image does not need to be a combination of the three spectrums. Instead, you process each image type through its respective model and ensemble the outputs post-detection.

Note that developing a system for this type of ensembling can be quite involved, as you need to implement a method to match boxes that correspond to the same object across the different model outputs. This can be done by implementing Intersection over Union (IoU) thresholds or other matching strategies to determine which boxes are referring to the same object in the scene.

Remember that testing and validation are crucial. You'll need to experiment to see what weights work best for combining the boxes and whether this ensembling technique indeed adds value in your specific application.

The ensemble model can potentially be more robust against various environmental conditions and times of day, leveraging the strengths of each type of sensor data. Keep iterating and validating the results to achieve the best performance for your use-case.

**Comment by Prem948** [comment]
> @jaideep11061982 to ensemble models in YOLOv8, you have to load each individual trained model and then pass an image or a set of images through each model one by one. Each model will output predictions which are basically bounding boxes with class scores.
> 
> To ensemble the predictions, you generally have two methods:
> 
> * **Voting-based method** (also known as majority voting): Each model votes for each class and the class with the most votes is chosen.
> * **Average method**: Each model outputs the probabilities for each class and the final output class is the one with the highest average probability.
> 
> Note that ensembling may not always lead to better performance. It highly depends on how diverse and accurate your individual models are. If your models are very similar or one model's performance is far worse than the others, ensembling might not help.
> 
> Always evaluate your ensemble model using a validation set to see if it actually improves the performance.
> 
> I hope this helps! Let me know if you have more questions.

Hi,

Can you provide some reference code for manual ensemble model in yolov8?

Thanks in advance

**Comment by Husnir24** [comment]
Hi Glenn Jocher, I want to ask your opinion if is it a good idea to ensemble three yolov8-pose models, in which each model had one pose as class, so there are three poses. I did this because when all three poses trained together in a model, the inference shows that the model did not perform very well.

**Comment by glenn-jocher** [comment]
@Husnir24 hi there! 🙌

Ensembling three YOLOv8-pose models, each trained on a specific pose, can be a clever approach, especially if you've observed that a single model struggles to distinguish between similar poses. By focusing each model on a distinct pose, you might enhance the detection accuracy for each specific pose.

Here's a simple way to ensemble your models in Python:

```python
from ultralytics import YOLO

# Load your trained models
model1 = YOLO('path/to/pose1_model.pt')
model2 = YOLO('path/to/pose2_model.pt')
model3 = YOLO('path/to/pose3_model.pt')

# Run inference on an image for each model
results1 = model1.predict('path/to/image.jpg')
results2 = model2.predict('path/to/image.jpg')
results3 = model3.predict('path/to/image.jpg')

# Combine results (example: taking the highest confidence score)
# This part depends on how you want to ensemble the results
combined_results = combine_results([results1, results2, results3])
```

Remember, the key to a successful ensemble is how you combine the results. Since your models are specialized, you might consider a strategy that leverages the confidence scores or even a voting mechanism.

Testing and validation are crucial to ensure this approach improves your model's performance. Good luck, and feel free to share your findings!

**Comment by ShyPooper** [comment]
> @Husnir24 hi there! 🙌
> 
> Ensembling three YOLOv8-pose models, each trained on a specific pose, can be a clever approach, especially if you've observed that a single model struggles to distinguish between similar poses. By focusing each model on a distinct pose, you might enhance the detection accuracy for each specific pose.
> 
> Here's a simple way to ensemble your models in Python:
> 
> ```python
> from ultralytics import YOLO
> 
> # Load your trained models
> model1 = YOLO('path/to/pose1_model.pt')
> model2 = YOLO('path/to/pose2_model.pt')
> model3 = YOLO('path/to/pose3_model.pt')
> 
> # Run inference on an image for each model
> results1 = model1.predict('path/to/image.jpg')
> results2 = model2.predict('path/to/image.jpg')
> results3 = model3.predict('path/to/image.jpg')
> 
> # Combine results (example: taking the highest confidence score)
> # This part depends on how you want to ensemble the results
> combined_results = combine_results([results1, results2, results3])
> ```
> 
> Remember, the key to a successful ensemble is how you combine the results. Since your models are specialized, you might consider a strategy that leverages the confidence scores or even a voting mechanism.
> 
> Testing and validation are crucial to ensure this approach improves your model's performance. Good luck, and feel free to share your findings!

Thanks for providing such an easy way to ensemble models. However, this is for prediction task. How should i do if I wanted to ensemble several models in validation task?

**Comment by tejasri19** [comment]
Hello @glenn-jocher, I want to perform ensemble of best variants of YOLO v5, YOLO v8, YOLO v9, YOLO v10 pn my custom dataset. I understand the logic given in above cell. But how to extract the label and score results of these models at inference stage. Please share code on how to apply logic to combine results.

**Comment by glenn-jocher** [comment]
Hi @tejasri19,

Great to hear that you're exploring model ensembling with YOLOv5, YOLOv8, YOLOv9, and YOLOv10! Combining the strengths of different models can indeed enhance performance. Below is an example of how you can extract labels and scores from each model and then combine the results:

```python
from ultralytics import YOLO

# Load your trained models
model_v5 = YOLO('path/to/yolov5_model.pt')
model_v8 = YOLO('path/to/yolov8_model.pt')
model_v9 = YOLO('path/to/yolov9_model.pt')
model_v10 = YOLO('path/to/yolov10_model.pt')

# Run inference on an image for each model
results_v5 = model_v5.predict('path/to/image.jpg')
results_v8 = model_v8.predict('path/to/image.jpg')
results_v9 = model_v9.predict('path/to/image.jpg')
results_v10 = model_v10.predict('path/to/image.jpg')

# Extract labels and scores
def extract_labels_and_scores(results):
    labels = results[0].boxes.cls.cpu().numpy()
    scores = results[0].boxes.conf.cpu().numpy()
    return labels, scores

labels_v5, scores_v5 = extract_labels_and_scores(results_v5)
labels_v8, scores_v8 = extract_labels_and_scores(results_v8)
labels_v9, scores_v9 = extract_labels_and_scores(results_v9)
labels_v10, scores_v10 = extract_labels_and_scores(results_v10)

# Combine results (example: weighted average based on confidence scores)
def combine_results(labels_list, scores_list):
    combined_labels = []
    combined_scores = []
    for i in range(len(labels_list[0])):
        combined_label = labels_list[0][i]  # Assuming same label across models
        combined_score = (scores_list[0][i] + scores_list[1][i] + scores_list[2][i] + scores_list[3][i]) / 4
        combined_labels.append(combined_label)
        combined_scores.append(combined_score)
    return combined_labels, combined_scores

combined_labels, combined_scores = combine_results([labels_v5, labels_v8, labels_v9, labels_v10],
                                                   [scores_v5, scores_v8, scores_v9, scores_v10])

print("Combined Labels:", combined_labels)
print("Combined Scores:", combined_scores)
```

This example demonstrates how to extract labels and scores from each model's inference results and then combine them using a simple averaging method. You can adjust the combination logic based on your specific requirements.

If you encounter any issues, please ensure you are using the latest versions of the models and libraries. For more detailed guidance, you can refer to our [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example) documentation.

Happy coding! 😊

**Comment by tejasri19** [comment]
Thank you @glenn-jocher. I understand that YOLO v8 deals with class imbalance. Is it the same with previous versions ( v3,v5,v7) and newer versions - v9 & v10. Do we need to apply nms on predictions from each model before combining them. I also understood that v10 is NMS free approach. Kindly share your thoughts.

**Comment by Ronvaknins** [comment]
Hi is it possible to use the tracking in a real-time video (slight delay) with multiple models ?
`from ultralytics import YOLO

modelA = YOLO('C:/DEV/ultralytics/c_models/A.pt')
modelB = YOLO('C:/DEV/ultralytics/c_models/B.pt')

#how to combine them?
results = model.track(source='rtsp://localhost:554/live', show=True, tracker="bytetrack.yaml")`

**Comment by pderrenger** [comment]
Yes, you can use multiple models for real-time video tracking with a slight delay. To combine the outputs of multiple models, you would need to run inference on each model separately and then merge the results. This involves processing the video stream frame-by-frame, applying each model, and then combining their predictions before displaying the results. While combining, you might need to handle overlapping detections and apply Non-Maximum Suppression (NMS) if necessary. For YOLOv10, which uses an NMS-free approach, you can directly merge the outputs.

**Comment by glenn-jocher** [comment]
@tejasri19 you're correct that YOLOv10 uses an NMS-free approach. For other versions like v3, v5, v7, v8, and v9, applying NMS on predictions before combining them is generally recommended to reduce duplicate detections. Each version has its own methods for handling class imbalance, so it's good to review their specific documentation for details.

**Comment by glenn-jocher** [comment]
@tejasri19 you're correct that YOLOv10 uses an NMS-free approach. For other versions like YOLOv3, v5, v7, v8, and v9, NMS is typically applied to filter predictions. When ensembling, it's generally a good idea to apply NMS to each model's predictions before combining them. This helps ensure cleaner and more accurate results.

**Comment by glenn-jocher** [comment]
You're correct that YOLOv10 uses an NMS-free approach. For other versions like v3, v5, v7, v9, applying NMS on predictions before combining them is typically necessary to reduce duplicate detections. Each version handles class imbalance differently, so it's good to check the specific documentation for details.

**Comment by glenn-jocher** [comment]
YOLOv8 and newer versions like YOLOv10 handle class imbalance more effectively than earlier versions. For ensembling models, applying NMS on predictions from each model before combining them is generally recommended, except for YOLOv10, which uses an NMS-free approach.

**Comment by tejasri19** [comment]
When yolov3, v5, v8, v9 have inherent NMS applied before giving final result, do I still have to run NMS when applying ensemble of the outputs of these models?
 @glenn-jocher @pderrenger

**Comment by glenn-jocher** [comment]
Yes, you should apply a final round of NMS when ensembling the outputs of multiple models, as individual NMS from each model will not address overlaps between their outputs. This ensures the combined results are refined and non-redundant.

**Comment by glenn-jocher** [comment]
To calculate mAP @john09282922 for an ensemble, process your dataset through the ensembled model (i.e., combine predictions from YOLOv9, v10, and v11 models as discussed earlier) and compare the predictions to your ground truth labels. Ultralytics does not provide built-in ensemble evaluation tools, so you'll need to write custom logic for combining predictions and calculating mAP using frameworks like [pycocotools library](https://pypi.org/project/pycocotools/). For reference on mAP calculation, you can review Ultralytics' implementation in their [metrics.py](https://github.com/ultralytics/ultralytics).
